src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs

```diff
@@ -284,13 +284,32 @@ private static void ExecutionContextCallback(object? s)
             public ExecutionContext? Context;
 
             /// <summary>A delegate to the <see cref="MoveNext()"/> method.</summary>
-            public Action MoveNextAction => _moveNextAction ??= new Action(MoveNext);
+            public Action MoveNextAction => _moveNextAction ??= GetMoveNextAction();
 
             internal sealed override void ExecuteFromThreadPool(Thread threadPoolThread) => MoveNext(threadPoolThread);
 
             /// <summary>Calls MoveNext on <see cref="StateMachine"/></summary>
             public void MoveNext() => MoveNext(threadPoolThread: null);
 
+            /// <summary> Creates a MoveNext action which is potentially wrapped to enable TPL etw and event source tracing. </summary>
+            private Action GetMoveNextAction()
+            {
+                Action moveNext = new Action(MoveNext);
+
+                // If logging is on, wrap the action so that the TPL event source can output 
+                // continuation info during the TaskWaitBegin event.
+                if (AsyncCausalityTracer.LoggingOn)
+                {
+                    moveNext = AsyncMethodBuilderCore.CreateContinuationWrapper(moveNext, (continuation, task) =>
+                    {
+                        // Simply call MoveNext();
+                        continuation();
+                    }, this);
+                }
+
+                return moveNext;
+            }
```

Comment:
@delmyers, can you please help me understand how this impacts the EventSource tracing?  I'm not seeing it.  CreateContinuationWrapper doesn't do any additional tracing.

Reply:
Sure, it is because of this code in TaskAwaiter.cs:

```CSharp
            if (log.IsEnabled())
            {
                // ETW event for Task Wait Begin
                Task? currentTaskAtBegin = Task.InternalCurrent;

                // If this task's continuation is another task, get it.
                Task? continuationTask = AsyncMethodBuilderCore.TryGetContinuationTask(continuation);
                log.TaskWaitBegin(
                    currentTaskAtBegin != null ? currentTaskAtBegin.m_taskScheduler!.Id : TaskScheduler.Default.Id,
                    currentTaskAtBegin != null ? currentTaskAtBegin.Id : 0,
                    task.Id, TplEventSource.TaskWaitBehavior.Asynchronous,
                    continuationTask != null ? continuationTask.Id : 0);
            }
```

``AsyncMethodBuilderCore.TryGetContinuationTask(continuation)`` attempts to cast the continuation as a ``ContinuationWrapper`` to grab its associated task. So, when we create the continuation action from ``AsyncStateMachineBox<TStateMachine>.MoveNextAction, we have to make sure that it is wrapped in a ``ContinuationWrapper`` in order to allow ``TaskAwaiter.OutputWaitEtwEvents(…)`` to grab the associated task.

Note the comment that I added below, though. ``TaskAwaiter.OutputWaitEtwEvents(…)`` checks for ``TplEventSource.Log.IsEnabled()``, whereas ``AsyncStateMachineBox<T>.GetMoveNextAction()`` checks for ``AsyncCausalityTracer.LoggingOn``. These aren't exactly symmetrical. However, ``AsyncCausalityTracer.LoggingOn`` was what was used prior to Core 3 for wrapping the continuation, and the logging that it means to record is the ``TraceSynchronousWorkStart/Completion``, so that is what I used as well.

Reply:
@stephentoub I've updated the comment to make it a little more clear.

Reply:
Thank you for the explanation.

I see now why your change addresses it.  However, it also adds non-trivial cost when logging is enabled, at least two allocations.  And while our logging-enabled path is more expensive than logging-disabled, we still want to try to minimize the expense, as we'd like developers to be able to enable logging with as little overhead as we can muster.

Given that, I'm not yet convinced it's the right fix.  For example, would it fix things to just change this:
https://github.com/dotnet/coreclr/blob/5b1c001bc0fb5bb1ce035c02ec275763f66defc8/src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs#L125-L126
to instead be this:
```C#
internal static Task? TryGetContinuationTask(Action continuation) =>
    continuation.Target is ContinuationWrapper wrapper ? wrapper._innerTask :
    continuation.Target as Task;
```
?

Reply:
I could give that a try as well. My first instinct was to use the same scheme that was used previously, which was to wrap the continuation. I'll check to make sure that there aren't any other side-effects of doing it this way.

Reply:
Thank you.

Reply:
@stephentoub, I've pushed the suggested changes to allow all of the automated tests to run. I'll let you know what my manual tests find as well.